File: brain_tools/migrate.js
// Migrate note frontmatter to the current schema version.

import { CURRENT_SCHEMA_VERSION } from "./schema.js";

export function needsMigration(meta) {
  return meta.schema_version !== CURRENT_SCHEMA_VERSION;
}

// Return a new metadata object upgraded to the current schema version.
export function migrateNote(meta) {
  const out = { ...meta };
  if (out.keywords === undefined) out.keywords = [];
  out.schema_version = CURRENT_SCHEMA_VERSION;
  return out;
}

File: src/main/java/brain/Migrate.java
package brain;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

/** Migrate note frontmatter to the current schema version. */
public final class Migrate {
  private Migrate() {}

  public static boolean needsMigration(Map<String, Object> meta) {
    return !Integer.valueOf(Schema.CURRENT_SCHEMA_VERSION).equals(meta.get("schema_version"));
  }

  /** Returns a new metadata map upgraded to the current schema version (input untouched). */
  public static Map<String, Object> migrateNote(Map<String, Object> meta) {
    Map<String, Object> out = new LinkedHashMap<>(meta);
    out.putIfAbsent("keywords", new ArrayList<>());
    out.put("schema_version", Schema.CURRENT_SCHEMA_VERSION);
    return out;
  }
}
